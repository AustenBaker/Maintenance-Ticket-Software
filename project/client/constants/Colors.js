const tintColor = '#2f95dc';

export default {
  tintColor,
  tabIconDefault: '#ccc',
  tabIconSelected: tintColor,
  tabBar: '#000',
  errorBackground: 'red',
  errorText: '#fff',
  warningBackground: '#EAEB5E',
  warningText: '#666804',
  noticeBackground: tintColor,
  noticeText: '#fff',
};
