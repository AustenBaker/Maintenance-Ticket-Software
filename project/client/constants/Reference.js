// REGEX constant 
export const REGEX = {
    // name pattern: up to 32 ASCII alphabetical characters
    NAME: /^[a-zA-Z]{1,32}$/,
    // email pattern:  some(.name)*@site.com
    // where some name & site are any combination of letters
    // and numbers, and the site ends in a 2-3 letter/digit extension
    EMAIL: /^[[\w]+[\.\w]*@[\w]+\.[\w]{2,3}]{6,32}$/,
    // phone pattern:  ###-###-#### where # is a digit 0-9
    PHONE: /^[\d]{3}-[\d]{3}-[\d]{4}$/,
    // password pattern: a-zA-Z0-9_-!@#$%^&*+=
    // requires at least one each: uppercase letter, number, punctuation
    // length of 8-32 characters
    PASSWORD: /^[[\w]*[A-Z]+[\d]+[\-!@#$%\^\*\+=]+]{8,32}$/
}


export const USER_TYPE = ['resident','maintenance','management'];
export const ENTRY_PERMISSION = ['accompanied','notify','any'];
export const PREFERRED_CONTACT = ['email','text'];

