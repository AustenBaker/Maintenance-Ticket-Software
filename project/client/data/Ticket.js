import * as React from 'react';
import { View, CheckBox, Picker, FormInput } from 'react-native';
import * as CONSTANTS from '../constants/Reference';

export default class Ticket extends React.Component{
  constructor(props) {
    super(props);
    this.state = {
      ticket_number: props.ticket_number,
      emergency: props.emergency,
      unit: props.unit,
      first_name: props.first_name,
      last_name: props.last_name,
      status: props.status,
      edit_mode: props.edit_mode,
      ticket_updates: [props.ticket_updates]
    }
  }

  /**
   * Declare default values for Ticket props
   */
  static defaultProps = {
    ticket_number: "99999",
    emergency: true,
    unit: "316",
    first_name: "FirstName",
    last_name: "LastName",
    status: "in-progress",
    edit_mode: "false",
    ticket_updates: []
  }

  //method to create ticket
  createTicket = () => {
	  
  }
  
  
  
  
  /**
   * This method displays tickets
   * emergency level, ticket number, unit number, status
   * display list of tickets
   * this should be called when a ticket from the list is pressed
   */
  displayTicketDetails = () => {
    var content;

    // create <Text> container for emergency data
    var emergency = "";
    if (this.emergency === true){
      emergency = (
        <Text>
          URGENT: This is an emergency level ticket!
        </Text>
      );
    }else{
      emergency = NULL;
    }

    //prepare content
    content = (
      <View>
        <Text>
          Ticket Number:
          {this.state.ticket_number}
          Unit Number:
          {this.state.unit}
          Status:
          {this.state.status}
        </Text>
        {emergency}
		//TODO add edit button
      </View>
    );
    return content;
  }
  
  //display ticket list using flatlist

  /**
   * This method updates ticket information
   * 
   * @param ticket_number Ticket number/id
   * @param emergency Ticket level
   * @param unit Ticket unit number
   * @param status Ticket status
   * 
   * @returns true if updated correctly
   */
  update = (props) => {
    var updated = false;
    var checked = [];
    if('ticket_number' in props){
      checked.push({['ticket_number']: props.ticket_number});
	}
    if('emergency' in props){
	  checked.push({['emergency']: props.emergency});
    }
    if('unit' in props){
	  checked.push({['unit']: props.unit});
    }
    if('status' in props){
	  checked.push({['status']: props.status});
    }

    //TODO: update server state and this.setState
    return updated;
  }

  /**
   * TODO: UPDATE
   * Generates form to display when editing a ticket
   * so far includes:
   * Ticket Number, Unit Number, Emergency/Ticket Level Checkbox,
   * Ticket Status Checkbox,  
   */
  editTicket = () => {
    var content;
    content = (
      <form name = "editTicket">
        <FormLabel>Ticket Number</FormLabel>
        <FormInput>
          onSubmitEditing={new_ticket_number => this.setState(ticket_number, new_ticket_number)}
          defaultValue={this.state.ticket_number}
        </FormInput>
        <FormValidationMessage>{'This field is required.'}</FormValidationMessage>

        <FormLabel>Unit Number</FormLabel>
        <FormInput>
          onSubmitEditing={newUnit => this.setState(unit, newUnit)}
          defaultValue={this.state.unit}
        </FormInput>
        <FormValidationMessage>{'This field is required.'}</FormValidationMessage>

        <FormLabel>Ticket Importance Level:</FormLabel>
        <Picker>
          <Picker
            title='Emergency'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={emergency === true}
            onIconPress={emergency = true}
          />
          <Picker.Item
            label='Normal'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={emergency === false}
            onIconPress={emergency = false}
          />
          <P
            label='Low'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={emergency === false}
            onIconPress={emergency = false}
          />
        </Picker>
        <FormValidationMessage>{'This field is required.'}</FormValidationMessage>

        <FormLabel>Ticket Status:</FormLabel>
        <Picker>
          <Picker.Item
            title='Open'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={status === 'open'}
            onIconPress={status = 'open'}
          />
          <Picker.Item
            title='In-Progress'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={status === 'in-progress'}
            onIconPress={status = 'in-progress'}
          />
          <Picker.Item
            title='Closed'
            checkedIcon='dot-circle-o'
            uncheckedIcon='circle-o'
            checked={status === 'closed'}
            onIconPress={status = 'closed'}
          />
        </Picker>
        <FormValidationMessage>{'This field is required.'}</FormValidationMessage>

      </form>
    );
    return content;
  }
  
  //create a close ticket method
  //method to delete tickets for management

  render = () => {
    var content;
    if(this.state.editMode === false){
      content = this.displayTicket();
    }else{
      content = this.editUser();
    }
    return (content);
  }

}