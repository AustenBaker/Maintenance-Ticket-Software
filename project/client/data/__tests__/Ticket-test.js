import * as React from 'react';
import * as JestNative from '@testing-library/jest-native';
import renderer from 'react-test-renderer';
import { getByPlaceholderText, render, getByTestId } from '@testing-library/react-native';
import Ticket from '../Ticket';
import * as CONSTANTS from '../../constants/Reference';

expect.extend({...JestNative})

jest.mock('expo', () => ({
    AppLoading: 'AppLoading',
}));

//renders ticket screen
test('renders display view correctly', () => {
    const tree = renderer.create(<Ticket />).toJSON();
    expect(tree).toMatchSnapshot();
});

//renders ticket edit screen
test('renders ticket edit view correctly', () => {
    const tree = renderer.create(<Ticket ticket_edit_mode />).toJSON();
    expect(tree).toMatchSnapshot();
});

describe(`The following are enabled from user perspective: `, () => {
  const { getByTestId } = render(<Ticket />);
  test('ticket_number', () => {
    expect(getByTestId(`ticket_number`)).toBeEnabled();
  });

  test('unit_number', () => {
    expect(getByTestId(`unit_number`)).toBeEnabled();
  });

  test('status', () => {
    expect(getByTestId(`status`)).toBeEnabled();
  });

  test('location', () => {
    expect(getByTestId(`location`)).toBeEnabled();
  });

  test('timestamp', () => {
    expect(getByTestId(`timestamp`)).toBeEnabled();
  });


});
