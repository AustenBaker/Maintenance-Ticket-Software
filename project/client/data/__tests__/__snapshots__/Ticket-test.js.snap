// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders display view correctly 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 12,
        "height": 120,
        "marginBottom": 20,
        "marginLeft": "10%",
        "marginTop": 20,
        "width": "80%",
      }
    }
    testID="ticketContainer"
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      Location: Apt. 
      316
      , 
      Whispering Pines
       
      

    </Text>
    <Text
      style={
        Object {
          "fontSize": 18,
        }
      }
    >
      <Text
        testID="status"
      >
        Status: Open
      </Text>
        |  
      <Text
        testID="emergency"
      >
        Level: Normal
      </Text>
        |
      

    </Text>
    <Text
      style={
        Object {
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      <Text
        testID="ticket_number"
      >
        Ticket ID #
        99999
      </Text>
    </Text>
    <Text
      style={
        Object {
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      Submitted: 
      12/31/1969 6:00:00 PM
    </Text>
  </View>
  <View
    style={
      Object {
        "alignSelf": "auto",
        "width": 400,
      }
    }
  >
    <View
      accessibilityLabel="Edit Ticket Button"
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Edit Ticket
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "auto",
        "width": 400,
      }
    }
  >
    <View
      accessibilityLabel="Edit Ticket Button"
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Show Ticket List
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders ticket edit view correctly 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 12,
        "height": 120,
        "marginBottom": 20,
        "marginLeft": "10%",
        "marginTop": 20,
        "width": "80%",
      }
    }
    testID="ticketContainer"
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      Location: Apt. 
      316
      , 
      Whispering Pines
       
      

    </Text>
    <Text
      style={
        Object {
          "fontSize": 18,
        }
      }
    >
      <Text
        testID="status"
      >
        Status: Open
      </Text>
        |  
      <Text
        testID="emergency"
      >
        Level: Normal
      </Text>
        |
      

    </Text>
    <Text
      style={
        Object {
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      <Text
        testID="ticket_number"
      >
        Ticket ID #
        99999
      </Text>
    </Text>
    <Text
      style={
        Object {
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      Submitted: 
      12/31/1969 6:00:00 PM
    </Text>
  </View>
  <View
    style={
      Object {
        "alignSelf": "auto",
        "width": 400,
      }
    }
  >
    <View
      accessibilityLabel="Edit Ticket Button"
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Edit Ticket
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "auto",
        "width": 400,
      }
    }
  >
    <View
      accessibilityLabel="Edit Ticket Button"
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Show Ticket List
        </Text>
      </View>
    </View>
  </View>
</View>
`;
